pkg/image/registryclient/client.go

```diff
@@ -199,7 +209,40 @@ func (c *Context) Ping(ctx context.Context, registry *url.URL, insecure bool) (h
 	return t, &src, nil
 }
 
-func (c *Context) Repository(ctx context.Context, registry *url.URL, repoName string, insecure bool) (distribution.Repository, error) {
+// Repository retrieves the appropriate repository implementation and ManifestService given a *Context
+func (c *Context) Repository(ctx context.Context, registry *url.URL, repoName string, insecure bool) (distribution.Repository, distribution.ManifestService, error) {
+	var repo distribution.Repository
+	var repoList []distribution.Repository
+	var err error
+	if len(c.ImageSources) == 0 {
+		repo, err = c.addPossibleRepository(ctx, registry, repoName, insecure)
+		if err != nil {
+			return nil, nil, err
+		}
+		repoList = append(repoList, repo)
+	}
+	var manifests distribution.ManifestService
+	for _, ics := range c.ImageSources {
+		repo, err = c.addPossibleRepository(ctx, ics.RegistryURL(), ics.RepositoryName(), insecure)
+		if err != nil {
+			return nil, nil, err
+		}
+		repoList = append(repoList, repo)
+	}
+	for _, r := range repoList {
+		manifests, err = repo.Manifests(context.TODO())
+		if err != nil {
+			err = fmt.Errorf("unable to get local manifest service: %v", err)
+			continue
+		}
+		return r, manifests, nil
```

Comment:
@sallyom sorry, I don't understand this PR at all. Let's focus on this code.

Creation a new network connection is expensive. Especially for disconnected clusters that drop packages to the external world. Let's imagine that you've mirrored `192.0.2.1/busybox` to `quay.io/prometheus/busybox` and you want to pull `192.0.2.1/busybox@sha256:64db0ecb0d989773dea64212d366e618c29a5552e3e386ac1dd603bf770f484e`. `192.0.2.1` is a dead address and normally is not available (https://tools.ietf.org/html/rfc5735, TEST-NET-1). `addPossibleRepository` will take few minutes for this address. Disconnected clusters see `quay.io` in a similar way: all traffic might be silently dropped and in this case `addPossibleRepository` will take few minutes before it returns an error.

Note that it's perfectly valid and expected that `quay.io` is not available on disconnected clusters.

And once you've created all these repositories, you take the first one, call `Manifests` (it will always succeed in our case), and return it. Everything else is thrown away.

Reply:
*upd:* `192.0.2.1` is a dead address, but `quay.io` is available. Therefore, if you have ICSP like

```
- mirrors:
  - quay.io/prometheus/busybox
  source: 192.0.2.1/busybox
```

pulling from `192.0.2.1/busybox` should work without any delays.

Reply:
I don't understand this scenario.  My use case for the change, and addressing https://bugzilla.redhat.com/show_bug.cgi?id=1823143 , is this:
* A release is mirrored from quay to a private disconnected registry
* A user wants to extract from that mirrored release

Currently this is not possible, because ICSPs are not considered, so _only_ the quay image is tried.  
This change adds mirrors from ICSP and returns the repository that is accessible given the Context.

Reply:
Part of this is a tradeoff - the client has different resiliency needs than a cluster.  If we *always* check ICSP in the client, then *every* image CLI command is slower.  I think unlike a cluster where we assume that the ICSP load is free (it's distributed up front to every node for resiliency), in the client we're biasing towards latency, so we want the client to avoid having to do the lazy fetch of ICSP unless it has to.  I think it's the right tradeoff.

Reply:
If we're making a client-specific tradeoff, I think it's best to have that functionality only in oc, since library-go is so widely used.

Reply:
@smarterclayton I see several options:

1. clients try to use ICSP only when they are explicitly asked to do it (via `--icsp-file` or `--use-cluster-icsp`), and they follow normal ICSP flow (mirrors first, then the original repo)
2. even if `--icsp-file` is provided, ICSP is loaded only after the initial attempt to access the original repo has failed (either from the cluster or from the file)
3. similar to the 2nd approach, but order of requests depends on the used flag (it's cheap to load ICSP from a file, it's more expensive to load it from a cluster)

You prefer the second option, right?

And no matter what option we choose, we shouldn't ping all registries in this method. We should try to use them one by one.

Reply:
we settled on this awhile ago:
* if --icsp-file is set, then decode/use that - don't look further
* image reference (this is the original image from original registry/repo) 
* if ICSP found in cluster, add that to list of image source options, those will be:
    * icsp from cluster
    * user-passed release image (substitute registry/repo/name from user-given w/ same digest)

My preference is that oc just works, it will look for image sources in that order without user specifying anything (except the --icsp-file)

Reply:
wrt @soltysh question, I can implement the RepositoryRetriever and keep these changes within oc, just have to add the RepositoryRetriever interface to Context and export a few methods.

Reply:
I've added the RepositoryRetriever interface to the registry client Context, see [the oc PR](https://github.com/openshift/oc/pull/439) for implementation.

Reply:
I'll try to explain what this code is doing again (it's in a different repo now, but changes in this repo are needed for it).

```go
	for _, ics := range c.ImageSources {
		repo, err = c.addPossibleRepository(ctx, ics.RegistryURL(), ics.RepositoryName(), insecure)
		if err != nil {
			return nil, nil, err
		}
		repoList = append(repoList, repo)
	}
```

For every ImageSource it'll send a GET request to something like `https://registry.example.com/v2/`. It may also try to send a request to `http://...` if the first request has failed and `insecure` is set. If the registry is not available, it will return an error.

I.e. for each ImageSource will have an instance of `Repository`. If at least one image source is not available, it'll be a fatal error.

```go
	for _, r := range repoList {
		manifests, err = repo.Manifests(context.TODO())
		if err != nil {
			err = fmt.Errorf("unable to get local manifest service: %v", err)
			continue
		}
		return r, manifests, nil
	}
```

`repoList` contains at least one element at this point, repo.Manifests [cannot fail](https://github.com/docker/distribution/blob/f4506b517a7e9e64cfdcc73c991d9e85b784dc3e/registry/client/repository.go#L177-L185), so the first element will be always returned. Other repoList elements will be thrown away.

ICSPs are created to allow clients to pull images when the source repository is not accessible. ICSPs can contain multiple mirrors so that if a single mirror is not available, the image is still pullable. This code assumes that all sources including all mirrors are available and this is against the idea of mirroring.

Reply:
Right, this should break when it returns a first repo. I've updated that in the oc PR, thank you. repoList is not necessary there, only ever dealing w/ a single repo 
```
	for _, ics := range c.ImageSources {
		repo, err = c.addPossibleRepository(ctx, ics.RegistryURL(), ics.RepositoryName(), insecure)
		if err != nil {
			return nil, nil, err
		}
		break
	}
	manifests, err = repo.Manifests(context.TODO())
	if err != nil {
		return nil, nil, fmt.Errorf("unable to get local manifest service: %v", err)
	}
	return repo, manifests, nil
```

@smarterclayton @dmage other than that, ok w/ this  PR?

Reply:
No, as I can see, `addPossibleRepository` is a copy of `(*registryclient.Context).Repository`. Instead of creating a copy and making helpers public, you can call the original function from`(c *registryContext).Repository` using `c.Context.Repository`. This function shouldn't return a ManifestService as the caller can do it itself. So these changes aren't necessary.

Reply:
The snippet isn't ok either, there is no guarantee that the first repository has necessary images. OpenShift uses ICSPs to generate registries.conf on nodes, from [containers-registries.conf(5)](https://github.com/containers/image/blob/master/docs/containers-registries.conf.5.md):

> The mirrors are attempted in the specified order; the first one that can be contacted and contains the image will be used (and if none of the mirrors contains the image, the primary location specified by the `registry.location` field, or using the unmodified user-specified reference, is tried last).

From [ICSP docs](https://github.com/openshift/api/blob/31aac52f4998f11617144a0d917c8ea1d2a1a9fc/operator/v1alpha1/types_image_content_source_policy.go#L59-L66):

> mirrors is one or more repositories that _may_ also contain the same images.

We've agreed that `oc` can use a different order, but it doesn't mean that `oc` can assume that all mirrors contain the same set of images.

Reply:
> No, as I can see, `addPossibleRepository` is a copy of `(*registryclient.Context).Repository`. Instead of creating a copy and making helpers public, you can call the original function from`(c *registryContext).Repository` using `c.Context.Repository`. This function shouldn't return a ManifestService as the caller can do it itself. So these changes aren't necessary.

I see that addPossibleRepository/making helpers public is not necessary, I'll update that now. Why not return ManifestService, since 99% of the time that's what we're after when looking for an appropriate Repository? The caller can easily call that but the duplicate code adds up so i don't see why we shouldn't include it.

Reply:
the manifestService returned guarantees the repo contains the image - I feel we're going in circles here.
If you don't agree we should return the ManifestService from the Repository fn, then I'll add that check in oc, in oc's RepositoryRetriever implementation.  However, I'll have to check for an error and throw out the manifestService returned from Repository(bc it won't return the ManifestService) but then I'll have to grab it elsewhere, as everywhere (almost) oc calls Repository, oc wants the ManifestService.  Is there a different way to ensure the Repository returned contains the image we're passing to find the Repository? If so  please let me know.

Reply:
met w/ @smarterclayton i have a different path forward to allow specific ordering of retrieving Manifests from a list of possible if/when that is appropriate (as opposed to lazy lookup/return otherwise) - i'll push for review once i sort out details

Reply:
@sallyom 

> the manifestService returned guarantees the repo contains the image

I gave you this link before, please take a look at the implementation of repo.Manifests, it cannot fail: https://github.com/docker/distribution/blob/f4506b517a7e9e64cfdcc73c991d9e85b784dc3e/registry/client/repository.go#L177-L185

ManifestService will always be created, its creation doesn't guarantee anything at all.

Reply:
I see, got it, the caller will still verify, then. It still makes sense to return the ManifestService with the Repository, IMO, since it saves a lot of duplicate code.

Reply:
removed the returning of the ManifestService, updated.